Block Library: Make it possible to import individual blocks

gziolo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Exploration to make it possible to import individual core blocks from `@wordpress/block-library`.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

At the moment, you can either import all core blocks with:

```js
import { registerCoreBlocks } from '@wordpress/block-library';

registerCoreBlocks();
```

If you would like to import individual blocks, you can still do it, but it isn't straightforward:

```js
import * as verse from '@wordpress/block-library/build-module/verse';

const { metadata, settings, name } = verse;
return registerBlockType( { name, ...metadata }, settings );
```

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

We could organize the `@wordpress/block-library` package to make the following possible:

```js
import { init } from '@wordpress/block-library/build-module/verse';

init();
```

or

```js
import '@wordpress/block-library/build-module/verse/init';
```

or

```js
import verseBlock from '@wordpress/block-library/build-module/verse/init';
```

### Implementation notes

My goal was to have a more ergonomic imports without `/build-module/` section like:

```js
import '@wordpress/block-library/verse/init';
```

However, it isn't that simple to achieve with partial ES Modules support with [`exports` field in the `package.json` file](https://nodejs.org/api/packages.html#modules-packages). We can change that in the future when we convert all WordPress packages to ES Modules.

Some blocks run some custom logic before registering the block so the newly introduced `init` method is a perfect way to postpone their executing until their are essential for the corresponding block to work correctly. This also removes the need to keep the list of block files with side-effects in `package.json.

gziolo committed:
Block Library: Make it possible to import individual blocks

sgomes reviewed (commented):
Thank you so much for working on this, @gziolo! 🙌 Sorry for the late reply 😞

In general it looks great, but I don't think the current code reflects the code example in the PR description. Taking the archives as the example, it seems to me that the correct code should instead be:

```js
import '@wordpress/block-library/archives/init';
```

if the user only cares about registering, or

```js
import archiveBlock from '@wordpress/block-library/archives/init';
```

if the user wants a reference to the block after it's registered.

The other option is of course to change the code to reflect the example instead, but having a separate `init.js` file seems like the better and most reliable (tree-shaking-wise) option to me.

In that case, though, `packages/block-library/src/archives/init.js` would need to be added to the list of modules with side effects in `package.json`, as it performs registration at the top level, which is a side effect.

gziolo commented:
@sgomes, I completed the refactoring based on your feedback. I also updated the description of the PR with the following examples:

```js
import { init } from '@wordpress/block-library/verse';

init();
```

or

```js
import '@wordpress/block-library/verse/init';
```

or

```js
import verseBlock from '@wordpress/block-library/verse/init';
```

I added a new section to the [README file](https://github.com/WordPress/gutenberg/blob/ccfd4c26ad067b25c51f71f450d0cdeb924b9b6d/packages/block-library/README.md) for the `@wordpress/block-library` package that covers it as well.

I'm happy about the idea I had where I created an exported `init` function for every block inside their `index.js` file. This way I could move all side effects to be evaluated only when that function gets called. In result, the list of side effects in `package.json` looks as follows:

```json
"sideEffects": [
	"build-style/**",
	"src/**/*.scss",
	"{src,build,build-module}/*/init.js"
],
```

The other thing I played with is `exports` in `package.json` so we don't have to use `build-module` for the import statements:

```json
"exports": {
	".": {
		"import": "./build-module/index.js",
		"require": "./build/index.js"
	},
	"./babel-plugin": "./babel-plugin.js",
	"./*": "./build-module/*/index.js",
	"./*/init": "./build-module/*/init.js",
	"./utils/*": null
},
```

I'm using it for the first time, so I'm not entirely sure if that is correct usage. It probably will be a breaking change for projects that use undocumented code like:

```js
import * as verse from '@wordpress/block-library/build-module/verse';
```

I kept CJS support for requiring only the `@wordpress/block-library`. It's something that we should get rid of at some point as well to make the size of the package smaller when publishing to npm.

Anyway, I'm looking for more feedback now that all the required changes are applied.

gziolo commented:
This is a bit surprising. I will investigate what went wrong.

<img width="741" alt="Screenshot 2022-07-22 at 14 10 11" src="https://user-images.githubusercontent.com/699132/180436390-4539777f-1dbc-46bf-a1bb-2955b184e159.png">

gziolo commented:
It looks like bringing back `module` in `package.json` with https://github.com/WordPress/gutenberg/pull/42258/commits/6a790d22029deb55556a9d182a66ec87226dc63a resolves the issue with the increased size of the `@wordpress/block-library` package. I will need to investigate how to update `package.json` correctly.

Unfortunately, having an additional `init` function in every core block has some tiny impact on the size of the entry point:

<img width="695" alt="Screenshot 2022-07-22 at 17 14 49" src="https://user-images.githubusercontent.com/699132/180470347-45346da5-c959-4fc5-b971-5fb3d9a58ca5.png">

sgomes commented:
> It looks like bringing back `module` in `package.json` with [6a790d2](https://github.com/WordPress/gutenberg/commit/6a790d22029deb55556a9d182a66ec87226dc63a) resolves the issue with the increased size of the `@wordpress/block-library` package. I will need to investigate how to update `package.json` correctly.

Interesting, thank you for looking into that! Note that this may be an artifact of how the GH action is building things, specifically the webpack config it uses. There are a lot of moving parts with webpack and package.json, unfortunately 😞

sgomes commented:
> Unfortunately, having an additional `init` function in every core block has some tiny impact on the size of the entry point:

That may still go away, depending on what the final `package.json` ends up being, but even if it stays I think it's worth the extra modularity 👍

sgomes commented:
As for the exports configuration, that one's pretty complex, with several things I hadn't seen before. It looks good, but I'm not quite sure how to test it for correctness 😕

gziolo commented:
I brought this branch up to date. I was playing a bit more with the `package.json` changes to test whether it's possible to use simplified import paths with ES Modules syntax. However, it's something we probably can't easily offer without converting all packages to ESM first (`"type": "modules"`), as we learned from https://github.com/WordPress/gutenberg/issues/43434. I tried to rename `init.js` files to ESM by renaming them to `init.mjs` but it can't be as simple as that because the code imports CJS which isn't allowed...

gziolo commented:
@sgomes, do you think it would still be valuable to include the changes and document the more verbose syntax that will work with the existing `package.json`:

```js
import verseBlock from '@wordpress/block-library/build-module/verse/init';
```

sgomes commented:
> @sgomes, do you think it would still be valuable to include the changes and document the more verbose syntax that will work with the existing `package.json`?

I think so, because it seems to me that having a (documented) mechanism to get proper tree-shaking, even if verbose, is better than not having that mechanism at all. Projects could start making use of it now, and then hopefully move to a terser import once everything has been converted to ESM.

Thanks again for working on this, @gziolo! 🙏

gziolo commented:
> I think so, because it seems to me that having a (documented) mechanism to get proper tree-shaking, even if verbose, is better than not having that mechanism at all. Projects could start making use of it now, and then hopefully move to a terser import once everything has been converted to ESM.

I went ahead and updated the branch to use the paths including `/build-module/` 👍🏻 

I think this is ready for the final review.

sgomes commented on packages/block-library/README.md:
````diff
@@ -90,6 +34,28 @@ _Parameters_
 
 <!-- END TOKEN(Autogenerated API docs) -->
 
+## Registering individual blocks
+
+1.  When you only care about registering the block when file gets imported:
+
+    ```js
+    import '@wordpress/block-library/build-module/verse/init';
+    ```
+
+2.  When you want to use the reference to the block after it gets automatically registered:
+
+    ```js
+    import verseBlock from '@wordpress/block-library/build-module/verse/init';
+    ```
+
+3.  When you need a full control over when the block gets registered:
+
+    ```js
+    import { init } from '@wordpress/block-library/build-module/verse';
+
+    const verseBlock = init();
+    ```
+
````
Perfect! Very clear and easy to follow 👍

sgomes commented on packages/block-library/README.md:
```diff
@@ -100,17 +66,57 @@ To find out more about contributing to this package or Gutenberg as a whole, ple
+
+    Then add `blinkingParagraph` to the list in the `getAllBlocks()` function.
+
+    If it's experimental, add the following property to `block. json`:
```
There's a stray space after `block.`.

gziolo replied:
Fixed 👍🏻

sgomes reviewed (approved):
LGTM from an API and documentation perspective. Thank you, @gziolo! 👍

gziolo commented on packages/block-library/README.md:
```diff
@@ -100,17 +66,57 @@ To find out more about contributing to this package or Gutenberg as a whole, ple
+
+    Then add `blinkingParagraph` to the list in the `getAllBlocks()` function.
+
+    If it's experimental, add the following property to `block. json`:
```
```suggestion
    If it's experimental, add the following property to `block.json`:
```

gziolo committed:
Update packages/block-library/README.md
